src/snowflake/snowpark/catalog.py

```diff
@@ -0,0 +1,659 @@
+                )
+            arg_types_str = ", ".join(
+                [convert_sp_to_sf_type(arg_type) for arg_type in arg_types]
+            )
+            return f"{fn}({arg_types_str})"
+
+        arg_types_str = ", ".join(arg.datatype for arg in fn.arguments)
+        return f"{fn.name}({arg_types_str})"
+
+    def _initialize_regex_udf(self) -> None:
+        with self._session._lock:
+            if self._python_regex_udf is not None:
+                return
+
+            def python_regex_filter(pattern: str, input: str) -> bool:
+                return bool(re.match(pattern, input))
+
+            self._python_regex_udf = self._session.udf.register(python_regex_filter)
+
+    def _list_objects(
+        self,
+        *,
+        object_name: str,
+        object_class,
+        database: Optional[Union[str, Database]],
+        schema: Optional[Union[str, Schema]],
+        pattern: Optional[str],
+    ):
+        db_name = self._parse_database(database)
+        schema_name = self._parse_schema(schema)
+
+        df = self._session.sql(
+            f"SHOW AS RESOURCE {object_name} IN {db_name}.{schema_name} -- catalog api"
+        )
+        if pattern:
+            # initialize udf
+            self._initialize_regex_udf()
+
+            # The result of SHOW AS RESOURCE query is a json string which contains
+            # key 'name' to store the name of the object. We parse json for the returned
+            # result and apply the filter on name.
+            df = df.filter(
+                self._python_regex_udf(
+                    lit(pattern), parse_json('"As Resource"')["name"]
+                )
+            )
+
+        return list(map(lambda row: object_class.from_json(row[0]), df.collect()))
+
+    # List methods
+    def list_databases(
+        self,
+        *,
+        pattern: Optional[str] = None,
+    ) -> List[Database]:
+        """List databases in the current session.
+
+        Args:
+            pattern: the pattern of name to match. Defaults to None.
+        """
+        iter = self._root.databases.iter()
+        if pattern:
+            iter = filter(lambda x: re.match(pattern, x.name), iter)
+
+        return list(iter)
+
+    def list_schemas(
+        self,
+        *,
+        database: Optional[Union[str, Database]] = None,
+        pattern: Optional[str] = None,
+    ) -> List[Schema]:
+        """List schemas in the current session. If database is provided, list schemas in the
+        database, otherwise list schemas in the current database.
+
+        Args:
+            database: database name or ``Database`` object. Defaults to None.
+            pattern: the pattern of name to match. Defaults to None.
+        """
+        db_name = self._parse_database(database)
+        iter = self._root.databases[db_name].schemas.iter()
+        if pattern:
+            iter = filter(lambda x: re.match(pattern, x.name), iter)
+        return list(iter)
```

Comment:
DatabaseModel and SchemaModel don't have from_json methods so we cannot use udf filter for them.